File: apps/frontend/src/lib/api/agents.ts
export interface ChatMessage {
  role: "user" | "agent";
  content: string;
  agent_name?: string;
}

export interface SessionHistory {
  session_id: string | null;
  messages: ChatMessage[];
}

export async function getSessionHistory(tripId: string): Promise<SessionHistory> {
  const res = await fetch(
    `http://localhost:8000/api/v1/agents/sessions/${tripId}`
  );
  if (!res.ok) return { session_id: null, messages: [] };
  return res.json();
}

export async function runAgent(
  tripId: string,
  message: string,
  sessionId: string | null,
  onToken: (token: string) => void,
  onDone: (sessionId: string, agentName: string) => void,
) {
  const res = await fetch("http://localhost:8000/api/v1/agents/run", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      trip_id: tripId,
      message,
      session_id: sessionId,
    }),
  });

  if (!res.ok) throw new Error("Agent request failed");

  const reader = res.body!.getReader();
  const decoder = new TextDecoder();
  let agentName = "Coordinator";

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    const chunk = decoder.decode(value);
    const lines = chunk.split("\n");

    for (const line of lines) {
      if (!line.startsWith("data: ")) continue;
      const data = line.slice(6);

      if (data === "[DONE]") continue;
      if (data.startsWith("[SESSION:")) {
        const id = data.slice(9, -1);
        onDone(id, agentName);
        continue;
      }
      if (data.startsWith("[AGENT:")) {
        agentName = data.slice(7, -1);
        continue;
      }
      onToken(data);
    }
  }
}
export async function saveAgentMessage(
  sessionId: string,
  agentName: string,
  content: string,
): Promise<void> {
  await fetch(
    `http://localhost:8000/api/v1/agents/sessions/${sessionId}/messages`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ agent_name: agentName, content }),
    }
  );
}

File: apps/backend/app/api/v1/routes/agents.py
import sys
from pathlib import Path

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.schemas.agent import AgentRunRequest
from app.models.trip import AgentSession, Message
from app.repositories.trip_repository import TripRepository

sys.path.insert(0, str(Path(__file__).parents[6] / "services" / "ai-agents"))

router = APIRouter(prefix="/agents", tags=["agents"])

NODE_NAME_MAP = {
    "destination_node": "Destination",
    "budget_node": "Budget",
    "hotel_node": "Hotel",
    "itinerary_node": "Itinerary",
    "coordinator_node": "Coordinator",
}


@router.post("/run")
async def run_agent(
    request: AgentRunRequest,
    db: AsyncSession = Depends(get_db),
):
    repo = TripRepository(db)
    trip = await repo.get_by_id(request.trip_id)
    if not trip:
        raise HTTPException(status_code=404, detail="Trip not found")

    if request.session_id:
        result = await db.execute(
            select(AgentSession).where(AgentSession.id == request.session_id)
        )
        session = result.scalar_one_or_none()
    else:
        session = None

    if not session:
        session = AgentSession(trip_id=trip.id)
        db.add(session)
        await db.flush()
        await db.refresh(session)

    user_message = Message(
        session_id=session.id,
        role="user",
        content=request.message,
    )
    db.add(user_message)
    await db.flush()

    session_id = session.id
    trip_data = {
        "destination": trip.destination,
        "duration_days": trip.duration_days,
        "budget": float(trip.budget),
        "currency": trip.currency,
        "preferences": trip.preferences or "",
        "trip_id": str(trip.id),
    }

    async def stream():
        from graph.workflow import graph
        from langchain_core.messages import HumanMessage

        state = {
            "messages": [HumanMessage(content=request.message)],
            "current_agent": "",
            "is_complete": False,
            "destination_research": "",
            "budget_breakdown": "",
            "hotel_suggestions": "",
            "itinerary_draft": "",
            "final_itinerary": "",
            **trip_data,
        }

        full_response = ""
        agent_name = "Coordinator"

        yield f"data: [AGENT:{agent_name}]\n\n"

        async for event in graph.astream_events(state, version="v2"):
            kind = event.get("event")

            if kind == "on_chain_start":
                node = event.get("name", "")
                if node in NODE_NAME_MAP:
                    agent_name = NODE_NAME_MAP[node]
                    yield f"data: [AGENT:{agent_name}]\n\n"

            if kind == "on_chat_model_stream":
                chunk = event["data"]["chunk"]
                token = chunk.content
                if token:
                    full_response += token
                    yield f"data: {token}\n\n"

        async with db.begin_nested():
            agent_msg = Message(
                session_id=session_id,
                role="agent",
                agent_name=agent_name,
                content=full_response,
            )
            db.add(agent_msg)

        yield f"data: [DONE]\n\n"
        yield f"data: [SESSION:{session_id}]\n\n"

    return StreamingResponse(
        stream(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "X-Accel-Buffering": "no",
        },
    )